File: Projects/Multithreading/app/src/main/java/com/ivettevaldez/multithreading/exercises/exercise5/ProducerConsumerBenchmarkUseCase.java
package com.ivettevaldez.multithreading.exercises.exercise5;

import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import com.ivettevaldez.multithreading.common.BaseObservable;

class ProducerConsumerBenchmarkUseCase extends BaseObservable<ProducerConsumerBenchmarkUseCase.Listener> {

    public interface Listener {

        void onBenchmarkCompleted(Result result);
    }

    static class Result {

        private final long executionTime;
        private final int numberOfMessages;

        Result(long executionTime, int numberOfMessages) {
            this.executionTime = executionTime;
            this.numberOfMessages = numberOfMessages;
        }

        long getExecutionTime() {
            return executionTime;
        }

        int getNumberOfMessages() {
            return numberOfMessages;
        }
    }

    private final static int NUM_OF_MESSAGES = 1000;
    private final static int BLOCKING_QUEUE_CAPACITY = 5;

    private final static Object LOCK = new Object();

    private final MyBlockingQueue blockingQueue = new MyBlockingQueue(BLOCKING_QUEUE_CAPACITY);
    private final Handler uiHandler = new Handler(Looper.getMainLooper());
    private final String TAG = this.getClass().getSimpleName();

    private int numOfReceivedMessages;
    private int numOfFinishedConsumers;
    private long startTimestamp;

    void startBenchmarkAndNotify() {
        synchronized (LOCK) {
            numOfReceivedMessages = 0;
            numOfFinishedConsumers = 0;
            startTimestamp = System.currentTimeMillis();
        }

        // Watcher-reporter thread.
        new Thread(() -> {
            synchronized (LOCK) {
                while (numOfFinishedConsumers < NUM_OF_MESSAGES) {
                    try {
                        LOCK.wait();
                    } catch (InterruptedException ex) {
                        Log.e(TAG, ex.toString());
                        return;
                    }
                }
            }

            notifySuccess();
        }).start();

        // Producers init thread.
        new Thread(() -> {
            for (int i = 0; i < NUM_OF_MESSAGES; i++) {
                startNewProducer(i);
            }
        }).start();

        // Consumers init thread.
        new Thread(() -> {
            for (int i = 0; i < NUM_OF_MESSAGES; i++) {
                startNewConsumer();
            }
        }).start();
    }

    private void notifySuccess() {
        uiHandler.post(() -> {
            Result result;

            synchronized (LOCK) {
                long executionTime = System.currentTimeMillis() - startTimestamp;
                result = new Result(executionTime, numOfReceivedMessages);
            }

            for (Listener listener : getListeners()) {
                listener.onBenchmarkCompleted(result);
            }
        });
    }

    private void startNewProducer(final int index) {
        new Thread(() -> blockingQueue.put(index)).start();
    }

    private void startNewConsumer() {
        new Thread(() -> {
            int message = blockingQueue.take();

            synchronized (LOCK) {
                if (message != -1) {
                    numOfReceivedMessages++;
                }
                numOfFinishedConsumers++;
                LOCK.notifyAll();
            }
        }).start();
    }
}

File: Projects/Multithreading/app/src/main/java/com/ivettevaldez/multithreading/demos/democoroutines/ProducerConsumerBenchmarkUseCase.kt
package com.ivettevaldez.multithreading.demos.democoroutines

import android.os.Handler
import android.os.Looper
import com.ivettevaldez.multithreading.common.BaseObservable
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

private const val NUM_OF_MESSAGES = 1000
private const val BLOCKING_QUEUE_CAPACITY = 5

class ProducerConsumerBenchmarkUseCase : BaseObservable<ProducerConsumerBenchmarkUseCase.Listener>() {

    private val reentrantLock = ReentrantLock()
    private val lockCondition = reentrantLock.newCondition()

    private val uiHandler = Handler(Looper.getMainLooper())
    private val blockingQueue = MyBlockingQueue(BLOCKING_QUEUE_CAPACITY)

    private var numOfReceivedMessages: Int = 0
    private var numOfFinishedConsumers: Int = 0
    private var startTimestamp: Long = 0

    interface Listener {

        fun onBenchmarkCompleted(result: Result?)
    }

    class Result(val executionTime: Long, val numberOfMessages: Int)

    fun startBenchmarkAndNotify() {
        reentrantLock.withLock {
            numOfReceivedMessages = 0
            numOfFinishedConsumers = 0
            startTimestamp = System.currentTimeMillis()
        }

        // Watcher/reporter thread.
        Thread {
            reentrantLock.withLock {
                while (numOfFinishedConsumers < NUM_OF_MESSAGES) {
                    try {
                        lockCondition.await()
                    } catch (ex: InterruptedException) {
                        return@Thread
                    }
                }
            }

            notifySuccess()
        }.start()

        // Producers thread.
        Thread {
            for (i in 0 until NUM_OF_MESSAGES) {
                startNewProducer(i)
            }
        }.start()

        // Consumers thread.
        Thread {
            for (i in 0 until NUM_OF_MESSAGES) {
                startNewConsumer()
            }
        }.start()
    }

    private fun notifySuccess() {
        uiHandler.post {
            lateinit var result: Result

            reentrantLock.withLock {
                result = Result(
                        System.currentTimeMillis() - startTimestamp,
                        numOfReceivedMessages
                )
            }
            
            for (listener in listeners) {
                listener.onBenchmarkCompleted(result)
            }
        }
    }

    private fun startNewProducer(i: Int) {
        Thread { blockingQueue.put(i) }.start()
    }

    private fun startNewConsumer() {
        Thread {
            val message = blockingQueue.take()

            reentrantLock.withLock {
                if (message != -1) {
                    numOfReceivedMessages++
                }
                numOfFinishedConsumers++
                lockCondition.signalAll()
            }
        }.start()
    }
}